src/common/type/pack.c

```diff
@@ -0,0 +1,1611 @@
+
+                // Read high order bits of the field ID delta when specified
+                if (tag & 0x2)
+                    this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 1;
+
+                // Read value
+                this->tagNextValue = (tag >> 2) & 0x3;
+            }
+        }
+        // Else the value is a single bit (e.g. boolean)
+        else if (packTypeData[this->tagNextType].valueSingleBit)
+        {
+            // Read low order bits of the field ID delta
+            this->tagNextId = tag & 0x3;
+
+            // Read high order bits of the field ID delta when specified
+            if (tag & 0x4)
+                this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 2;
+
+            // Read value
+            this->tagNextValue = (tag >> 3) & 0x1;
+        }
+        // Else the value is multiple tags (e.g. container)
+        else
+        {
+            // Read low order bits of the field ID delta
+            this->tagNextId = tag & 0x7;
+
+            // Read high order bits of the field ID delta when specified
+            if (tag & 0x8)
+                this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 3;
+
+            // Value length is variable so is stored after the tag
+            this->tagNextValue = 0;
+        }
+
+        // Increment the next tag id
+        this->tagNextId += this->tagStackTop->idLast + 1;
+
+        // Tag was found
+        result = true;
+    }
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+/***********************************************************************************************************************************
+Read field tag
+
+Some tags and data may be skipped based on the value of the id parameter.
+***********************************************************************************************************************************/
+static uint64_t
+pckReadTag(PackRead *this, unsigned int *id, PackType type, bool peek)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM_P(UINT, id);
+        FUNCTION_TEST_PARAM(ENUM, type);
+        FUNCTION_TEST_PARAM(BOOL, peek);                            // Look at the next tag without advancing the field id
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+    ASSERT(id != NULL);
+    ASSERT((peek && type == pckTypeUnknown) || (!peek && type != pckTypeUnknown));
+
+    // Increment the id by one if no id was specified
+    if (*id == 0)
+    {
+        *id = this->tagStackTop->idLast + 1;
+    }
+    // Else check that the id has been incremented
+    else if (*id <= this->tagStackTop->idLast)
+        THROW_FMT(FormatError, "field %u was already read", *id);
+
+    // Search for the requested id
+    do
+    {
+        // Get the next tag if it has not been read yet
+        if (this->tagNextId == 0)
+            pckReadTagNext(this);
+
+        // Return if the id does not exist
+        if (*id < this->tagNextId)
```

Comment:
I'm a little confused here - before, if the ID did not exist then we'd error if they weren't just trying to "peek" at it. So now that we're breaking out of the loop if it doesn't exist, then what are we returning with this->tagNextValue?

Reply:
This is because we no longer error if you try to read something and it does not exist (when defaultNull = false before). Now we just always default it so there's no way to get an error.

this->tagNextValue will be the value of the next existing tag -- if we get to this condition it means the thing we are looking for is before it and therefore NULL.